File: frontend/src/app/services/club-member.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpResponse} from "@angular/common/http";
import {Observable} from "rxjs";
import {ClubMember} from "../model/clubMember";


@Injectable({
  providedIn: 'root'
})
/**
 * The implementation of a service that cancel, create, update and delete club members
 *
 * @author Kim Bartikowski, Ertan Arabaci, Torben Labs
 *
 */
export class ClubMemberService {

  constructor(private http: HttpClient) {
  }

  /**
   * creates a club member
   *
   * @param clubMember an empty club member
   * @return ClubMember
   */
  createClubMember(clubMember: ClubMember) {
    return this.http.post<ClubMember>('/rest/club-members', clubMember)
  }

  /**
   * loads all club members that are saved
   *
   * @return a list of all club members
   */
  loadClubMembers(): Observable<ClubMember[]> {
    return this.http.get<ClubMember[]>('/rest/club-members');
  }

  /**
   * gets an empty club member
   *
   * @return an empty club member
   */
  getEmptyClubMember(): Observable<ClubMember> {
    return this.http.get<ClubMember>('/rest/empty-club-member')
  }

  /**
   * get specific club member by id
   *
   * @param id the id of the club member that is to be returned
   * @return the club member with given id
   */
  getClubMember(id: number): Observable<ClubMember> {
    return this.http.get<ClubMember>('/rest/club-member-form/' + id);
  }

  /**
   * gets club member by id for the accounting overview
   *
   * @param id the id of the club member that is to be returned
   * @return the club member with given id
   */
  getClubMemberAccounting(id: number): Observable<ClubMember> {
    return this.http.get<ClubMember>('/rest/club-member-accounting/' + id);
  }

  /**
   * updates the attributes of a given club member
   *
   * @param clubMember the club member that is to be updated
   * @return the updated club member
   */
  updateClubMember(clubMember: ClubMember): Observable<ClubMember> {
    return this.http.put<ClubMember>(`/rest/club-members`, clubMember)
  }

  /**
   * deletes a club member by given id
   *
   * @param id id of the club member that is to be deleted
   * @return a Http response based on the club member id
   */
  deleteClubMember(id: number): Observable<HttpResponse<void>> {
    return this.http.delete<void>(`/rest/club-members/${id}`, {observe: 'response'});
  }

}

File: backend/src/main/java/de/nordakademie/iaa/clubmembermgt/service/ClubMemberService.java
package de.nordakademie.iaa.clubmembermgt.service;

import de.nordakademie.iaa.clubmembermgt.dao.ClubMemberDAO;
import de.nordakademie.iaa.clubmembermgt.model.ClubMember;
import org.hibernate.exception.ConstraintViolationException;

import javax.inject.Inject;
import java.util.List;

/**
 * The implementation of a service for ClubMember
 *
 * @author Kim Bartikowski, Torben Labs, Ertan Arabaci
 */
public class ClubMemberService {

    private ClubMemberDAO clubMemberDAO;

    public void createClubMember(ClubMember clubMember) {
        clubMember.setExitDate(null);
        clubMember.setTerminationDate(null);
        clubMemberDAO.createClubMember(clubMember);
    }


    public void updateClubMember(ClubMember clubMember) {
        clubMemberDAO.updateClubMember(clubMember);
    }

    public void deleteClubMember(Long id) {
        clubMemberDAO.deleteClubMember(id);
    }

    public ClubMember loadEmptyClubMember() {
        return clubMemberDAO.loadEmptyClubMember();
    }


    @SuppressWarnings("unchecked")
    public List<ClubMember> listClubMembers() {
        return clubMemberDAO.listClubMembers();
    }

    public ClubMember loadClubMember(Long clubMemberId) {
        return clubMemberDAO.loadClubMember(clubMemberId);
    }

    @Inject
    public void setClubMemberDAO(ClubMemberDAO clubMemberDAO) {
        this.clubMemberDAO = clubMemberDAO;
    }
}

